massa-factory-worker/src/denunciation_factory.rs

```diff
@@ -0,0 +1,303 @@
+
+        // Get selected address from selector and check
+        let selected = self.channels.selector.get_selection(endo_slot);
+        match selected {
+            Ok(selection) => {
+                if let Some(address) = selection.endorsements.get(endo_index_usize) {
+                    if *address
+                        != Address::from_public_key(
+                            &secure_share_endorsement.content_creator_pub_key,
+                        )
+                    {
+                        warn!("Denunciation factory received a secure share endorsement but address was not selected");
+                        return;
+                    }
+                } else {
+                    warn!("Denunciation factory could not get selected address for endorsements at index: {}", endo_index_usize);
+                    return;
+                }
+            }
+            Err(e) => {
+                warn!("Cannot get producer from selector: {}", e);
+            }
+        }
+
+        if self.endorsements_by_slot_index.len() > DENUNCIATION_FACTORY_ENDORSEMENT_CACHE_MAX_LEN {
+            warn!(
+                "Denunciation factory cannot process - cache full: {}",
+                self.endorsements_by_slot_index.len()
+            );
+            return;
+        }
+
+        let denunciation_: Option<Denunciation> = match self.endorsements_by_slot_index.entry(key) {
+            Entry::Occupied(mut eo) => {
+                match eo.get_mut() {
+                    EndorsementDenunciationStatus::Accumulating(secure_endo_1_) => {
+                        let secure_endo_1: &SecureShareEndorsement = secure_endo_1_;
+                        match Denunciation::try_from((secure_endo_1, &secure_share_endorsement)) {
+                            Ok(de) => {
+                                eo.insert(EndorsementDenunciationStatus::DenunciationEmitted);
+                                Some(de)
+                            }
+                            Err(e) => {
+                                debug!("Denunciation factory cannot create denunciation from endorsements: {}", e);
+                                None
+                            }
+                        }
+                    }
+                    EndorsementDenunciationStatus::DenunciationEmitted => {
+                        // A Denunciation has already been created, nothing to do here
+                        None
+                    }
+                }
+            }
+            Entry::Vacant(ev) => {
+                ev.insert(EndorsementDenunciationStatus::Accumulating(
+                    secure_share_endorsement,
+                ));
+                None
+            }
+        };
+
+        if let Some(denunciation) = denunciation_ {
+            info!(
+                "Created a new endorsement denunciation : {:?}",
+                denunciation
+            );
+            self.channels.pool.add_denunciation(denunciation);
+        }
+
```

Comment:
Same simplification should also apply here

```suggestion
match self.endorsements_by_slot_index.entry(key) {
    Entry::Occupied(mut eo) => {
        match eo.get_mut() {
            EndorsementDenunciationStatus::Accumulating(secure_endo_1_) => {
                let secure_endo_1: &SecureShareEndorsement = secure_endo_1_;
                match Denunciation::try_from((secure_endo_1, &secure_share_endorsement)) {
                    Ok(de) => {
                        eo.insert(EndorsementDenunciationStatus::DenunciationEmitted);
                        info!("Created a new endorsement denunciation : {:?}", de);
                        self.channels.pool.add_denunciation(de);
                    }
                    Err(e) => {
                        debug!("Denunciation factory cannot create denunciation from endorsements: {}", e);
                    }
                }
            }
            EndorsementDenunciationStatus::DenunciationEmitted => {}
        }
    }
    Entry::Vacant(ev) => {
        ev.insert(EndorsementDenunciationStatus::Accumulating(
            secure_share_endorsement,
        ));
    }
}

```